src/main/java/com/palantir/gradle/versions/CheckOverbroadConstraints.java

```diff
@@ -0,0 +1,165 @@
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.gradle.api.DefaultTask;
+import org.gradle.api.file.RegularFile;
+import org.gradle.api.file.RegularFileProperty;
+import org.gradle.api.provider.Property;
+import org.gradle.api.tasks.Input;
+import org.gradle.api.tasks.InputFile;
+import org.gradle.api.tasks.Internal;
+import org.gradle.api.tasks.TaskAction;
+import org.gradle.api.tasks.options.Option;
+import org.gradle.language.base.plugins.LifecycleBasePlugin;
+
+public abstract class CheckOverbroadConstraints extends DefaultTask {
+
+    @Internal
+    protected abstract Property<Boolean> getShouldFixProperty();
+
+    @Internal
+    protected abstract RegularFileProperty getPropsFileProperty();
+
+    @Internal
+    protected abstract RegularFileProperty getLockFileProperty();
+
+    public CheckOverbroadConstraints() {
+        getShouldFixProperty().set(false);
+        setGroup(LifecycleBasePlugin.VERIFICATION_GROUP);
+        setDescription(
+                "Ensures matched versions in your versions.lock are pinned to avoid wasted dependency resolution.");
+        getOutputs().upToDateWhen(_task -> true); // task has no outputs, this is needed for it to be up to date
+    }
+
+    @InputFile
+    public final Property<RegularFile> getLockFile() {
+        return getLockFileProperty();
+    }
+
+    @InputFile
+    public final Property<RegularFile> getPropsFile() {
+        return getPropsFileProperty();
+    }
+
+    @Input
+    public final Property<Boolean> getShouldFix() {
+        return getShouldFixProperty();
+    }
+
+    @Option(option = "fix", description = "Whether to apply the suggested fix to versions.props")
+    public final void setShouldFix(boolean shouldFix) {
+        this.getShouldFixProperty().set(shouldFix);
+    }
+
+    @TaskAction
+    public final void checkOverbroadConstraints() {
+        VersionsProps versionsProps =
+                VersionsProps.loadFromFile(getPropsFile().get().getAsFile().toPath());
+        LockState lockState =
+                new ConflictSafeLockFile(getLockFile().get().getAsFile().toPath()).readLocks();
+
+        checkOverbroadConstraints(versionsProps, lockState);
+    }
+
+    private void checkOverbroadConstraints(VersionsProps versionsProps, LockState lockState) {
+
+        List<String> newLines = determineNewLines(versionsProps, lockState);
+
+        if (newLines.isEmpty()) {
+            return;
+        } else if (getShouldFixProperty().get()) {
+            getLogger()
+                    .lifecycle("Adding pins to versions.props:\n"
+                            + newLines.stream().collect(Collectors.joining("\n")));
+            writeVersionsProps(getPropsFile().get().getAsFile(), newLines);
+            return;
+        }
+
+        throw new RuntimeException("There are efficient pins missing from your versions.props: \n"
```

Comment:
Need a better exception message than this. Maybe we need a good one line description of the problem and a readme section we can link to. There's a good explanation of the problem [here](https://pl.ntr/2oW) that we could probably take.

Reply:
I've added to the readme and improve the exception message